src/effects/effectslot.cpp

```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+    for (int parameterTypeId=0 ; parameterTypeId<numTypes ; ++parameterTypeId) {
+        const EffectManifestParameter::ParameterType parameterType =
+                static_cast<EffectManifestParameter::ParameterType> (parameterTypeId);
+
+        m_pControlNumParameters[parameterType]->forceSet(numParameters(parameterType));
+
+        unsigned int iParameter = 0;
+        unsigned int numParameterSlots = m_iNumParameterSlots[parameterType];
+        for (int i=0 ; i<m_parameterSlotPositionToManifestIndex[parameterType].size() ; ++i) {
+            const int iParameterPosition = m_parameterSlotPositionToManifestIndex[parameterType].value(i, -1);
+
+            auto pParameter = m_parameters.value(iParameterPosition, nullptr);
+
+            VERIFY_OR_DEBUG_ASSERT(pParameter != nullptr) {
+                continue;
+            }
+
+            if (iParameter < numParameterSlots) {
+                auto pParameterSlot = getEffectParameterSlot(parameterType, iParameter);
+                VERIFY_OR_DEBUG_ASSERT(pParameterSlot != nullptr) {
+                    break;
+                }
+                pParameterSlot->loadParameter(pParameter);
+                ++iParameter;
+            }
+        }
+        while (iParameter < numParameterSlots) {
+            auto pParameterSlot = getEffectParameterSlot(parameterType, iParameter);
+            VERIFY_OR_DEBUG_ASSERT(pParameterSlot != nullptr) {
+                break;
+            }
+            pParameterSlot->clear();
+            ++iParameter;
+        }
     }
+}
+
+void EffectSlot::unloadEffect() {
+    if (!isLoaded()) {
+        return;
+    }
+
     m_pControlLoaded->forceSet(0.0);
-    m_pControlNumParameters->forceSet(0.0);
-    m_pControlNumButtonParameters->forceSet(0.0);
-    for (const auto& pParameter : m_parameters) {
-        pParameter->clear();
+    for (const auto& pControlNumParameters : m_pControlNumParameters) {
+        pControlNumParameters->forceSet(0.0);
     }
-    for (const auto& pParameter : m_buttonParameters) {
-        pParameter->clear();
+
+    for (const auto& pParameterSlot : m_parameterSlots) {
+        pParameterSlot->clear();
+    }
+
+    for (int i = 0; i < m_parameters.size(); ++i) {
+        EffectParameter* pParameter = m_parameters.at(i);
+        m_parameters[i] = nullptr;
+        delete pParameter;
+    }
+    m_parameters.clear();
+    m_pManifest.clear();
+
+    removeFromEngine();
+}
+
+void EffectSlot::hideEffectParameter(const unsigned int parameterId) {
+    for (auto& parameterMapping : m_parameterSlotPositionToManifestIndex) {
+        parameterMapping.removeAll(parameterId);
+    }
+
+    loadParameters();
+}
+
+void EffectSlot::setEffectParameterPosition(const unsigned int parameterId,
+        const unsigned int position) {
+    for (auto& parameterMapping : m_parameterSlotPositionToManifestIndex) {
+        parameterMapping.removeAll(parameterId);
+    }
+
+    auto pParameter = m_parameters.at(parameterId);
```

Comment:
I'm not sure we can depend on parameters being listed with all knobs then all buttons. For Mixxx's built in effects, we could easily document it should be that way. But I don't think we can rely on that for external effects plugins.

Reply:
I think we should rework this function to be more robust or somehow enforce the constraints it on EffectManifests that it assumes.

Reply:
Maybe we can reorder the effect parameters according to the parameter type after loading an `EffectManifest` and overwrite the same?

Reply:
I'm leaning towards refactoring this function to take the EffectManifestParameter::ParameterType as an argument. We already store a separate map for each parameter type.

Currently this function is only used by DlgPrefLV2 (indirectly via EffectsManager and EffectChainSlot) for hiding LV2 parameters. DlgPrefLV2 does not distinguish between types of parameters and making it do that would add more complexity to the code. I think at this point we should go ahead and remove DlgPrefLV2. I think it would be more worthwhile to get started working towards saving custom per-effect default snapshots with XML than continue maintaining the hacky way hiding LV2 parameters was implemented before.